JDBI Module, Binders currently duplicated in our services

freddd opened the pull request:
As we discussed before we should have a JDBI lib as these classes are generic and can be used by whoever, wherever (JDBI is used of course). These are basically copies of what's currently in accounting-service & translation-service. I had to make one small changes because I did not want to add guava as a dependency just to use Lists & Iterables. Also worth mentioning is that this module is overriding the default source & target, Java 7 and is instead using Java 8. It's a quick change if we would want this to be compiled using 7 instead.

The binders are used to automagic IN-clauses (*ArrayBinder) or change value to bind on (enums is a good example where the default is name() but sometimes we want ordinal()):
* ```EnumOrdinalBinder```
* ```EnumNameArrayBinder```
* ```IntegerArrayBinder```
* ```StringArrayBinder```

The ```PostgresArrayArgumentFactory``` is the ArgumentFactory that will intercept and do the actual binding.
The ```SqlArray``` captures the information to be bound.

Anyway, let me know what you think.

ping @aaanders @goncalosmo, @fredrikbackstrom, this might also be of interest for @adamvoncorswant as it goes into the toolbox which is OS.

adamvoncorswant commented on izettle-jdbi/pom.xml:
```diff
@@ -0,0 +1,41 @@
+            <groupId>org.jdbi</groupId>
+            <artifactId>jdbi</artifactId>
+            <version>${jdbi.version}</version>
+            <scope>provided</scope>
```
Does this mean that the dependency will be downloaded during build, but not included in the generated artifact?

freddd replied:
Yes, exactly. It will not be packaged in the izettle-jdbi jar but expects the container/jdk to have the dependency already. I assume that if someone is using the izettle-jdbi they also have the jdbi-jar already in their container, though I guess one could argue that they might use a different version but then we cannot guarantee that our extensions work as intended.

nzroller replied:
Actually I think you'll run into issue with using `provided` here. The generally usage of this is with dropwizard apps and we shade all dependencies into the one uber jar. So there is no container and the jdk doesn't have it. I don't think it is the usual usage of this scope.

adamvoncorswant commented on izettle-jdbi/src/main/java/com/izettle/jdbi/PostgresArrayArgumentFactory.java:
```diff
@@ -0,0 +1,39 @@
+        Object[] elements = value.getElements();
+        if (elements != null && elements.length > 0) {
+            if (elements[0] instanceof Number) {
+                return "integer";
```
Is this really good enough? I mean, there are many types of numbers, such as `short, int, float, double`. Won't jdbi need to make the distinction?

freddd replied:
I agree, we should change this. Did a quick check in the accounting service, we're creating an SqlArray of Longs which we are then here returning as type "integer". I'll look into it tomorrow.

adamvoncorswant replied:
Check, thanks

adamvoncorswant commented:
I like this, and think it's the perfect place to put stuff like this.

fredrikbackstrom commented:
Nice things! @AndreasMeisingseth @tbaylis we might have things in risk repo that could go in here or at least copy-pastad here for use by others.

freddd commented:
Renamed the IntegerArrayBinder and added more granular clauses as discussed above. I will look into creating test cases using a dao so that we test the proper usage.

tbaylis commented:
Cool, once this is merged, I can look through which changes we can make in our code, and move some of our tools here aswell.

adamvoncorswant commented on izettle-jdbi/src/main/java/com/izettle/jdbi/PostgresArrayArgumentFactory.java:
```diff
@@ -0,0 +1,45 @@
+
+    public Argument build(Class<?> expectedType, final SqlArray<?> value, StatementContext ctx) {
+        return (position, statement, ctx1) -> {
+            final Array ary = ctx.getConnection().createArrayOf(getTypeName(value), value.getElements());
```
As this is OS, I think we should take a bit more pride in variable naming. What about `elements`?

freddd replied:
Yes. I renamed it to ```elements```.

adamvoncorswant commented on izettle-jdbi/src/main/java/com/izettle/jdbi/EnumNameArrayBinder.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public void bind(SQLStatement<?> q, Bind bind, Collection<Enum> arg) {
+        Set<String> names =
```
`final`?

adamvoncorswant commented on izettle-jdbi/src/main/java/com/izettle/jdbi/EnumNameArrayBinder.java:
```diff
@@ -0,0 +1,22 @@
+            arg
+                .stream()
+                .map(Enum::name)
+                .collect(Collectors.toSet());
```
Why not the above 5 lines on one line?

freddd replied:
As I said in other PR's we need to take a look at the formatting for Java 8. It seems our formatter is doing crazy stuff sometimes.

goncalosmo replied:
This one is my fault, though.

adamvoncorswant commented on izettle-jdbi/src/main/java/com/izettle/jdbi/PostgresArrayArgumentFactory.java:
```diff
@@ -0,0 +1,45 @@
+        return value instanceof SqlArray && clazz.isAssignableFrom(((SqlArray) value).getType());
+    }
+
+    public Argument build(Class<?> expectedType, final SqlArray<?> value, StatementContext ctx) {
```
Love that the `value` is `final`, but why not make all arguments `final` in this entire PR?

adamvoncorswant commented on izettle-jdbi/src/main/java/com/izettle/jdbi/SqlArray.java:
```diff
@@ -0,0 +1,36 @@
+
+        elements.forEach(listFromIterable::add);
+
+        return listFromIterable;
```
Just:
```java
return StreamSupport.stream(iterable.spliterator(), false).collect(Collectors.toList());
```
?

adamvoncorswant commented on izettle-jdbi/src/main/java/com/izettle/jdbi/SqlArray.java:
```diff
@@ -0,0 +1,36 @@
+        return new SqlArray<>(type, getListFromIterable(elements));
+    }
+
+    private static <T> List<T> getListFromIterable(Iterable<T> elements) {
```
Naming: Just `toList`?

adamvoncorswant commented on izettle-jdbi/src/test/java/com/izettle/jdbi/PostgresArrayArgumentSpec.java:
```diff
@@ -0,0 +1,44 @@
+        assertThat(LONG_FACTORY.getTypeName(new SqlArray<>(String.class, getList("test")))).isEqualTo("text");
+    }
+
+    private <T> List<T> getList(T value) {
```
Nitpicking naming here: this method name sounds like it's calling some unknown internal function on the `value` argument. Why not just name it `asList` instead?

adamvoncorswant commented on izettle-jdbi/src/test/java/com/izettle/jdbi/PostgresArrayArgumentSpec.java:
```diff
@@ -0,0 +1,44 @@
+        List<T> list = new ArrayList<>();
+        list.add(value);
+
+        return list;
```
Why not just:
```java
return Stream.of(value).collect(Collectors.toList());
```

freddd commented:
Added .DS_Store to the .gitignore and also added individual versioning on izettle-jdbi, I don't really see a reason as to why it would have the same as the toolbox as a whole. Anyway, please look through it again and see if it's ready to be merged.

nzroller commented on izettle-jdbi/pom.xml:
```diff
@@ -0,0 +1,42 @@
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>izettle-jdbi</artifactId>
+    <version>1.0.0-SNAPSHOT</version>
```
Version should stay the same as the toolbox proper I think (i.e. remove this line). For now it makes releases a tad easier.

freddd replied:
Alright, I'll change it. It's nice to not having it dependent on the toolbox though, as every time we make a change in a submodule we need to bump and release the toolbox even though it's quite rarely (if ever?) used as it's own entity as opposed to the submodules being used individually.

nzroller commented on izettle-jdbi/src/main/java/com/izettle/jdbi/EnumNameArrayBinder.java:
```diff
@@ -0,0 +1,18 @@
+        final Set<String> names = arg.stream().map(Enum::name).collect(Collectors.toSet());
+        q.bind(bind.value(), SqlArray.arrayOf(String.class, names));
+    }
+}
```
So this is truly minor but github is telling me something is wrong with not having a newline at the end of the file! I think git complains about that too... If you ever wondered why, the history is here: http://stackoverflow.com/questions/729692/why-should-files-end-with-a-newline and the first three answers all seem reasonable.

nzroller commented:
:+1:

freddd commented:
@nzroller I just pushed a couple of commits that removes the individual versioning and adds new lines to two that were missing it. Please have a look again.

xiaodong-izettle commented on izettle-jdbi/src/test/java/com/izettle/jdbi/PostgresArrayArgumentSpec.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    private <T> List<T> toList(T value) {
+        return Stream.of(value).collect(Collectors.toList());
```
Just use `Arrays.asList()`?

freddd commented:
Ready for merge?

nzroller commented:
Might want to do one last rebase here. Also `provided` scope is still there which I think will cause problems when we don't explicitly specify the jdbi dependency (for example when we get it as a transitive dependency from DW). I'm testing it to see though...

nzroller commented:
(woops I didn't rebase myself)...

freddd commented:
Alright, let me know regarding the provided scope. I generally think it's quite nice to have it as it makes our libs quite small but perhaps as you say, it will cause issues.

nzroller commented:
I double-checked and it actually didn't cause an issue. But I don't think we can make the assumption that it will be provided. In dropwizard apps no jar's are provided by the JDK and it's all shaded into one jar. If there is a situation where you need it provided I think it's best to then define the dependency and scope in that particular project.
